r1cs/gadgets/std/src/fields/nonnative/reduce.rs

```diff
@@ -252,25 +421,95 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
             ));
         }
 
-        // This part we mostly use the techniques in bellman-bignat
-        // The following code is adapted from https://github.com/alex-ozdemir/bellman-bignat/blob/master/src/mp/bignat.rs#L567
+        // The equality of two `A`-ary representations `[L[0],L[1],..]` and `[R[0],R[1],..]` with 
+        // oversized limbs is proven by enforcing their *shifted* differences 
+        // ``
+        //      [L[0] + (- R[0] + shift_constant) , 
+        //          L[1] + (- R[1] + shift_constant), 
+        //              ..],
+        // `` 
+        // to represent 
+        // ``
+        //      Sum_{i>=0} shift_constant * A^i.
+        // ``
+        // The constant `shift_constant = 2^bits_per_group  - 1` is to circumvent underflows in a
+        // length-preserving manner. With this choice  
+        // ``
+        //     0 <=  shift_constant - R[i] < 2^bits_per_group,
+        // `` 
+        // and hence
+        // ``
+        //      L[i] + (shift_constant - R[i]) < 2^{bits_per_group + 1},
+        // ``
+        // Since the length of the carries are throughout `<= bits_per_limb + surfeit`, 
+        // we also have `carry[i-1] + L[i] < 2^bits_per_group`, and the overall sum 
+        // ``
+        //      carry[i-1] + L[i] + (shift_constant - R[i]) 
+        // `` 
+        // is still of length `<= bits_per_group + 1`.
```

Comment:
I am not sure about that. Below a more detailed explanation.

Reply:
The fact that the carry doesn't cause an additional increase of the length is not obvious. Let's discuss it offline if my comments on this do not help.